In [1]:
import numpy as np
import pandas as pd
from scipy import stats

# Домашнее задание 8   

## Задача 1

> Даны значения величины заработной платы заемщиков банка (salary) и значения их поведенческого кредитного скоринга (scoring):  
> salary = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]  
> scoring = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]  
>
>- Найдите ковариацию этих двух величин: сначала без использования специальных функций, а затем с помощью функции numpy.cov. Полученные значения должны быть равны.  
>- Найдите коэффициент корреляции Пирсона: с помощью ковариации и средних квадратических отклонений двух признаков, а затем с помощью специальных функций из библиотек numpy и pandas.

1. Найдите ковариацию этих двух величин  
    
а. Без использования специальных функций
 

In [2]:
zp  = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

cov_r = (zp * ks).mean() - zp.mean() * ks.mean()

cov_r_s = ((zp - zp.mean()) * (ks - ks.mean())).sum() / (zp.shape[0] - 1)
print("Несмещенная        Смещенная\n",cov_r, cov_r_s)


Несмещенная        Смещенная
 9157.839999999997 10175.37777777778


    
б. С помощью функции numpy.cov

In [3]:
cov = np.cov(zp, ks, ddof=0)
cov_s = np.cov(zp, ks)
print("Несмещенная Смещенная")
cov[1, 0], cov_s[1, 0]

Несмещенная Смещенная


(9157.84, 10175.377777777776)

2) Найдите коэффициент корреляции 

а. Без использования специальных функций

In [4]:
corr = cov_r / (zp.std() * ks.std())
corr

0.8874900920739158

б. с помощью специальных функций из библиотек numpy и pandas

In [5]:
df = pd.DataFrame({'Salary': zp, 'Scoring': ks})
df[['Salary', 'Scoring']].corr()


Salary  Scoring
Salary   1.00000  0.88749
Scoring  0.88749  1.00000

In [6]:
np.corrcoef(zp, ks)[1, 0]

0.8874900920739162

## Задача 2

> Измерены значения IQ выборки студентов, обучающихся в местных технических вузах:  
> 131, 125, 115, 122, 131, 115, 107, 99, 125, 111  
> Известно, что в генеральной совокупности IQ распределен нормально.  
>- Найдите доверительный интервал для математического ожидания с надежностью 0.95.

Решение:

In [7]:
samples = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])

In [8]:
n = samples.shape[0]
m = samples.mean()
s = samples.std(ddof=1)
print(f'n = {n},среднее = {m}, оценка среднего квадратического отклонения  = {s}')

n = 10,среднее = 118.1, оценка среднего квадратического отклонения  = 10.54566788359614


In [9]:
t = stats.t.ppf(0.975, 9)
t

2.2621571627409915

Таким образом находим доверительный интервал для оценки истинного (генерального) значения:

In [10]:
print(m - t * s / np.sqrt(n),  m + t * s / np.sqrt(n))
stats.t.interval(0.95 , df=len(samples)-1, loc=np.mean(samples), scale=stats.sem(samples))

110.55608365158724 125.64391634841274


(110.55608365158724, 125.64391634841274)

Доверительный интервал:  
$\color{blue}(110.556; 125.644)$

## Задача 3.

> Известно, что рост футболистов в сборной распределен нормально с известной дисперсией 25. На выборке объёма 27 выборочное среднее составило 174.2.  
>
>- Найдите доверительный интервал для математического ожидания с надежностью 0.95.


$\underline{X}\pm Z_\frac{\alpha}{2}\cdot\frac{\sigma}{\sqrt{n}}$

$Z_\frac{\alpha}{2}=1.96$

$174.2\pm 1.96\cdot\frac{5}{\sqrt{27}}$


In [11]:
n = 27
m = 174.2
s = 5
z = 1.96
print(m - z * s / np.sqrt(n), m + z * s / np.sqrt(n))

172.31398912064722 176.08601087935276


$Ответ: \color{blue}[172.314;176.086]$